### PR TITLE
*: check data corruption on boot

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/8313.

### CODE REVIEW

File: e2e/ctl_v3_alarm_test.go [hunk 1]
```diff
@@ -99,6 +108,101 @@ func alarmTest(cx ctlCtx) {
 	}
 }
 
+type fakeConsistentIndex struct{ rev uint64 }
+
+func (f *fakeConsistentIndex) ConsistentIndex() uint64 { return f.rev }
+
+func alarmCorruptTest(cx ctlCtx) {
```
Comment (heyitsanthony) [inline e2e/ctl_v3_alarm_test.go]: why isn't this an integration test? there's already a corruption test there

Comment (gyuho) [reply e2e/ctl_v3_alarm_test.go]: If we fatal in corruption, fatal logger calls `os.Exit(1)` (so not testable)?

File: e2e/ctl_v3_alarm_test.go [hunk 2]
```diff
@@ -99,6 +108,101 @@ func alarmTest(cx ctlCtx) {
...
+			corrupted = true
+			break
+		}
+		time.Sleep(time.Second)
+	}
+	if !corrupted {
+		cx.t.Fatalf("expected error %v after %s", rpctypes.ErrCorrupt, 5*time.Second)
+	}
+
+	// corrupt alarm should now be on
```
Comment (heyitsanthony) [inline e2e/ctl_v3_alarm_test.go]: The cluster that's currently running can be expected to be OK since it's periodically checking its hash with its other members. The newly corrupted member should fatal out before joining raft / doing any damage / causing need for an alarm.

File: e2e/ctl_v3_alarm_test.go [hunk 3]
```diff
@@ -99,6 +108,101 @@ func alarmTest(cx ctlCtx) {
...
+
+	sresp, err := cli1.Status(context.TODO(), eps[0])
+	if err != nil {
+		cx.t.Fatal(err)
+	}
+	id0 := sresp.Header.GetMemberId()
+
+	corrupted := false
+	for i := 0; i < 5; i++ {
+		presp, perr := cli0.Put(context.TODO(), "abc", "aaa")
```
Comment (heyitsanthony) [inline e2e/ctl_v3_alarm_test.go]: the boot check should prevent member 0 from servicing any KV RPCs, otherwise it doesn't provide any better guarantees than the periodic corruption check

File: e2e/ctl_v3_alarm_test.go (hunk 2 above)
Comment (gyuho) [reply e2e/ctl_v3_alarm_test.go]: > The newly corrupted member should fatal out before joining raft

> Instead, if a member can contact the client addresses of its peers, it should first fetch hashes from the other members at a known revision and compare before serving any client requests. (from original issue)

Agree.

So should the checking be before `s.publish` request, in `s.start`, using Raft entries?

https://github.com/coreos/etcd/blob/master/etcdserver/server.go

```go
func (s *EtcdServer) Start() {
	s.start()
	s.goAttach(func() { s.publish(s.Cfg.ReqTimeout()) })
	s.goAttach(s.purgeFile)
	s.goAttach(func() { monitorFileDescriptor(s.stopping) })
	s.goAttach(s.monitorVersions)
	s.goAttach(s.linearizableReadLoop)
	s.goAttach(s.monitorKVHash)
}
```

I am trying to figure out the best way to fetch hashes from other members without starting gRPC server, or starting Raft. Or on restart, we still start Raft and gRPC server but do not accept new requests other than hash kv RPC?
> The newly corrupted member should fatal out before joining raft

> Instead, if a member can contact the client addresses of its peers, it should first fetch hashes from the other members at a known revision and compare before serving any client requests. (from original issue)

Agree.

So should the checking be before `s.publish` request, in `s.start`, using Raft entries?

https://github.com/coreos/etcd/blob/master/etcdserver/server.go

```go
func (s *EtcdServer) Start() {
	s.start()
	s.goAttach(func() { s.publish(s.Cfg.ReqTimeout()) })
	s.goAttach(s.purgeFile)
	s.goAttach(func() { monitorFileDescriptor(s.stopping) })
	s.goAttach(s.monitorVersions)
	s.goAttach(s.linearizableReadLoop)
	s.goAttach(s.monitorKVHash)
}
```

I am trying to figure out the best way to fetch hashes from other members without starting gRPC server, or starting Raft. Or on restart, we still start Raft and gRPC server but do not accept new requests other than hash kv RPC?

Comment (heyitsanthony) [reply e2e/ctl_v3_alarm_test.go]: Probably just before `s.start()` entirely; load the cluster membership info and hash from the backend, issue the hashkv RPCs to the member addresses and compare with the backend, then call start() if there's no mismatch.

**Comment by gyuho** [comment]
> Probably just before s.start() entirely; load the cluster membership info and hash from the backend, issue the hashkv RPCs to the member addresses and compare with the backend, then call start() if there's no mismatch.


> before s.start() entirely; load the cluster membership info

`start` has to be called first, to get Raft started. If there's no snapshot on restart, member lists are fetched via `RaftCluster.Members`. And `RaftCluster.Members` is populated via `EtcdServer.publish`, which requires Raft applier.

> and hash from the backend

This is possible if we start the peer listener first, as in this PR. Current master branch calls `EtcdServer.Start` first, before starting peer listeners (including gRPC peer handlers).

So now start peer gRPC listeners for Hash API, start Raft, `EtcdServer.publish`, and then check-initial-hash. But now `TestDetectKvOrderViolation` fails. Investigating...

**Comment by gyuho** [comment]
CI's failing from recent health balancer patches... Need fix them first.

**Comment by xiang90** [comment]
if this PR itself is done, can we remove the WIP label?

File: etcdserver/corrupt.go [hunk 4]
```diff
@@ -24,6 +24,69 @@ import (
...
+	}
+	mbs := s.cluster.Members()
+	n := len(mbs)
+	if n < 3 {
+		plog.Infof("no need to check corruption on %s (%d members)", s.ID().String(), n)
+		return
+	}
+	plog.Infof("corruption checking on %s (%d members)", s.ID().String(), n)
+
+	h, rev, _, err := s.kv.HashByRev(0)
```
Comment (xiang90) [inline etcdserver/corrupt.go]: what happens when this local node is slow, and the revision has been compacted on other peers?

File: etcdserver/corrupt.go [hunk 5]
```diff
@@ -24,6 +24,69 @@ import (
 	"github.com/coreos/etcd/pkg/types"
 )
 
+func (s *EtcdServer) checkHashKVInit() {
+	// TODO: separate configuration for initial hash check?
+	if s.Cfg.CorruptCheckTime < 3*time.Second {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: why?

File: etcdserver/corrupt.go [hunk 6]
```diff
@@ -24,6 +24,69 @@ import (
...
+	}
+
+	select {
+	case <-s.readych: // sync with s.publish
+	case <-s.stopping:
+		return
+	}
+	mbs := s.cluster.Members()
+	n := len(mbs)
+	if n < 3 {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: why?

File: etcdserver/corrupt.go [hunk 7]
```diff
@@ -24,6 +24,69 @@ import (
...
+			resp, cerr = cli.HashKV(ctx, c, rev)
+			cancel()
+			if cerr == nil {
+				break
+			}
+		}
+		cli.Close()
+
+		if resp == nil && cerr != nil {
+			plog.Fatal(cerr)
```
Comment (xiang90) [inline etcdserver/corrupt.go]: if all its peers are dead, then this peer cannot start. the cluster cannot be bootstrapped after a full shutdown?

**Comment by gyuho** [comment]
@xiang90 PTAL.

I've added `--experimental-initial-corrupt-check` flag in later commits, since `integration.TestV3CorruptAlarm` with `initial-corrupt-check` would fatal, which cannot be tested in integration tests. Also I think it's reasonable to have separate flags, since hash monitoring warns through the alarm, while initial hash check just fatals out the process even before serving any client/peer traffic.

File: etcdserver/corrupt.go [hunk 8]
```diff
@@ -19,11 +19,63 @@ import (
...
 )
 
+// checkInitialHashKV compares initial hash values with its peers
+// before serving any client traffic.
+func (s *EtcdServer) checkInitialHashKV() {
+	if !s.Cfg.InitialCorruptCheck {
+		return
+	}
+
+	// on restart, s.publish is required for member fetch
```
Comment (xiang90) [inline etcdserver/corrupt.go]: not sure what the comment means. need to make it more clear.

File: etcdserver/corrupt.go [hunk 9]
```diff
@@ -19,11 +19,63 @@ import (
...
+	select {
+	case <-s.readych:
+	case <-s.stopping:
+		return
+	}
+
+	plog.Infof("initial corruption checking on %s", s.ID())
```
Comment (xiang90) [inline etcdserver/corrupt.go]: etcdserver %d starting initial corruption check...

File: etcdserver/corrupt.go [hunk 10]
```diff
@@ -19,11 +19,63 @@ import (
...
+					mismatch++
+				}
+				// if this local node is slow while other peers already compacted
+				// on requested revision, then it's not corrupt; just this node is slow
+			}
+			continue
+		}
+		peerID := types.ID(p.resp.Header.MemberId)
+		if h != p.resp.Hash && rev == p.resp.Header.Revision && crev == p.resp.CompactRevision {
+			plog.Warningf("%s's hash %d != %s's hash %d (revision %d)", s.ID(), h, peerID, p.resp.Hash, rev)
```
Comment (xiang90) [inline etcdserver/corrupt.go]: this should be an error.

File: etcdserver/corrupt.go [hunk 11]
```diff
@@ -19,11 +19,63 @@ import (
...
+		}
+	}
+
+	quorum := len(s.cluster.Members())/2 + 1
+	// mismatch >= quorum, this node is corrupt; fatal to terminate
+	// mismatch < quorum, other peers are corrupt; nothing to do
+	if mismatch >= quorum {
+		plog.Fatalf("%s is corrupt", s.ID())
+	}
+	plog.Infof("initial corruption checking success on %s", s.ID())
```
Comment (xiang90) [inline etcdserver/corrupt.go]: %d succeeded on initial corruption checking: no corruption.

**Comment by xiang90** [comment]
@gyuho 

can we move the refactor and code structure changes to another PR. we can merge those commits before we merge the corruption related one.

**Comment by gyuho** [comment]
@xiang90 PTAL.

File: etcdserver/server.go [hunk 12]
```diff
@@ -192,6 +192,11 @@ type EtcdServer struct {
 	// when there is no error
 	readNotifier *notifier
 
+	// cmu protects channels below
```
Comment (xiang90) [inline etcdserver/server.go]: is this related to the corruption detection change?

File: etcdserver/server.go [hunk 13]
```diff
@@ -1039,7 +1047,12 @@ func (s *EtcdServer) stopWithDelay(d time.Duration, err error) {
 
 // StopNotify returns a channel that receives a empty struct
 // when the server is stopped.
-func (s *EtcdServer) StopNotify() <-chan struct{} { return s.done }
+func (s *EtcdServer) StopNotify() <-chan struct{} {
```
Comment (xiang90) [inline etcdserver/server.go]: we probably should also comment that stopNotify should only be called after start returns. or stopNotify can be pending forever. if we follow that convention, will race still happen?

Comment (gyuho) [reply etcdserver/server.go]: > stopNotify should only be called after start returns.

Now that `servePeers` start first, `v3rpc.newStreamInterceptor` with `v3rpc.monitorLeader` gets called before `e.Server.Start`, so there was race waiting on `done` channel.

> stopNotify can be pending forever if it gets the old done chan

it will timeout after election timeout

https://github.com/coreos/etcd/blob/15bfc1b36162805e8a90ae747d955667870c2f95/etcdserver/api/v3rpc/interceptor.go#L110-L118

> will race still happen

I think we still need locks because it will always be called before `e.Server.Start`?

File: etcdserver/server.go (hunk 12 above)
Comment (gyuho) [reply etcdserver/server.go]: Yes, since channel initialization now happens after `v3rpc.monitorLeader` started.

1. `servePeers`
2. `v3rpc.newStreamInterceptor` with `v3rpc.monitorLeader`
3. `e.Server.Start` that initializes the channels

File: etcdserver/corrupt.go [hunk 14]
```diff
@@ -125,7 +183,12 @@ func (s *EtcdServer) checkHashKV() error {
 	return nil
 }
 
-func (s *EtcdServer) getPeerHashKVs(rev int64) (resps []*clientv3.HashKVResponse) {
+type peerHashKV struct {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: peerHashKVResp

File: etcdserver/corrupt.go [hunk 15]
```diff
@@ -19,11 +19,65 @@ import (
...
+	plog.Infof("%s starting initial corruption check...", s.ID())
+	h, rev, crev, err := s.kv.HashByRev(0)
+	if err != nil {
+		plog.Fatalf("%s failed to fetch hash (%v)", s.ID(), err)
+	}
+
+	peers := s.getPeerHashKVs(rev)
+	mismatch := 0
+	for _, p := range peers {
+		if p.resp == nil {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: probably handle p.resp != nil case first. it is simpler and is the most common case. it improves readability.

File: etcdserver/corrupt.go [hunk 16]
```diff
@@ -19,11 +19,65 @@ import (
...
+			if p.err != nil {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: is there is a case that resp is nil, and err is nil?

File: etcdserver/corrupt.go [hunk 17]
```diff
@@ -19,11 +19,65 @@ import (
...
+					// then peer will return "mvcc: required revision is a future revision"
+					mismatch++
+				}
+				// if this local node is slow (e.g. ErrCompacted) on requested revision,
+				// then it's not corrupt; just this node is slow
+			}
+			continue
+		}
+		peerID := types.ID(p.resp.Header.MemberId)
+		if h != p.resp.Hash && rev == p.resp.Header.Revision && crev == p.resp.CompactRevision {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: why do we need to compare rev and crev? we request hash at rev. it if fails, error should be reported. we should trust the server response.

File: etcdserver/corrupt.go [hunk 18]
```diff
@@ -19,11 +19,65 @@ import (
...
+			plog.Errorf("%s's hash %d != %s's hash %d (revision %d)", s.ID(), h, peerID, p.resp.Hash, rev)
+			mismatch++
+		}
+	}
+
+	quorum := len(s.cluster.Members())/2 + 1
```
Comment (xiang90) [inline etcdserver/corrupt.go]: the node should refuse to start even if a single mismatch is detected. or the error might be propagated to the local node.

File: etcdserver/server.go (hunk 13 above)
Comment (xiang90) [reply etcdserver/server.go]: > Now that servePeers start first,

Why do we need this change?

File: etcdserver/corrupt.go (hunk 18 above)
Comment (gyuho) [reply etcdserver/corrupt.go]: What if

```
A: hash 15 (corrupt)
B: hash 10
C: hash 10
```

A restarts and fatal since it's corrupt.

B restarts at the same time as A and before A fatals out, then B mismatch count will be 1 while B is not corrupt?

Comment (xiang90) [reply etcdserver/corrupt.go]: Both A and B will fatal. Human operator needs to check what is going on there. We expect people to come to check the cluster when there is a hash mismatch no matter on minor nodes or majority nodes.

**Comment by gyuho** [comment]
@xiang90 PTAL. I simplified it all.

File: embed/etcd.go [hunk 19]
```diff
@@ -123,12 +123,15 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
...
 	)
-
+	urlsmap, token, err = cfg.PeerURLsMapAndToken("etcd")
 	if !isMemberInitialized(cfg) {
-		urlsmap, token, err = cfg.PeerURLsMapAndToken("etcd")
 		if err != nil {
 			return e, fmt.Errorf("error setting up initial cluster: %v", err)
 		}
+	} else {
+		// if member is already initialized,
```
Comment (xiang90) [inline embed/etcd.go]: if member is initialized, we should get the peer URLs from on disk configuration, not from cfg (which is from command line flags i assume?)

File: etcdserver/corrupt.go [hunk 20]
```diff
@@ -19,11 +19,49 @@ import (
...
+				plog.Errorf("%s's hash %d != %s's hash %d (revision %d)", s.ID(), h, peerID, p.resp.Hash, rev)
+				mismatch++
+			}
+			continue
+		}
+		if p.err != nil {
+			if p.err == rpctypes.ErrFutureRev {
+				plog.Errorf("%s has future revision %d (%q)", s.ID(), rev, p.err)
+			}
+			// p.err == rpctypes.ErrCompacted, then local node is slow (maybe not corrupt)
```
Comment (xiang90) [inline etcdserver/corrupt.go]: log this?

File: etcdserver/corrupt.go [hunk 21]
```diff
@@ -19,11 +19,49 @@ import (
...
+			peerID := types.ID(p.resp.Header.MemberId)
+			if h != p.resp.Hash && rev == p.resp.Header.Revision {
+				plog.Errorf("%s's hash %d != %s's hash %d (revision %d)", s.ID(), h, peerID, p.resp.Hash, rev)
+				mismatch++
+			}
+			continue
+		}
+		if p.err != nil {
+			if p.err == rpctypes.ErrFutureRev {
+				plog.Errorf("%s has future revision %d (%q)", s.ID(), rev, p.err)
```
Comment (xiang90) [inline etcdserver/corrupt.go]: cannot check the hash of peer %d: peer is lagging behind.

File: embed/etcd.go (hunk 19 above)
Comment (gyuho) [reply embed/etcd.go]: Yes, that's what `PeerURLsMapAndToken` is doing? This is just resetting `token` and `err` as before, since it used to be set only when `!isMemberInitialized(cfg)`

**Comment by gyuho** [comment]
@xiang90 PTAL. Now it loads peer URLs from WAL entries.

File: etcdserver/corrupt.go [hunk 22]
```diff
@@ -125,36 +170,59 @@ func (s *EtcdServer) checkHashKV() error {
...
+func (s *EtcdServer) getPeerHashKVs(rev int64) (resps []*peerHashKVResp) {
+	mbs := s.cluster.Members()
+	pURLs := make([][]string, len(mbs))
+	for _, m := range mbs {
 		if m.ID == s.ID() {
 			continue
 		}
+		pURLs = append(pURLs, m.PeerURLs)
+	}
+	if len(pURLs) == 0 {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: this is not needed. we never try to check hash on fresh node.

Comment (gyuho) [reply etcdserver/corrupt.go]: Found that `s.cluster.Members()` is only initially populated where there's any snapshot on disk

https://github.com/coreos/etcd/blob/c983f0ae988504159c6337761027aa5ef7a3fc48/etcdserver/server.go#L389-L396

So we need a way to fallback on Raft entries from disk when we read WAL.

File: embed/etcd.go [hunk 23]
```diff
@@ -186,6 +188,16 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
...
 	e.errc = make(chan error, len(e.Peers)+len(e.Clients)+2*len(e.sctxs))
 
+	// newly started member ("memberInitialized==false")
+	// does not need corruption check
+	if memberInitialized {
+		if err = e.Server.CheckInitialHashKV(); err != nil {
+			// since this is before "EtcdServer.Start()"
+			// if not nil, it will block on "EtcdServer.Close()"
+			e.Server = nil
+			return e, err
```
Comment (xiang90) [inline embed/etcd.go]: just return nil, err

File: etcdserver/corrupt.go [hunk 24]
```diff
@@ -16,14 +16,56 @@ package etcdserver
...
+			switch p.err {
+			case rpctypes.ErrFutureRev:
+				plog.Errorf("%s cannot check the hash of peer(%q) at revision %d: peer is lagging behind(%q)", s.ID(), p.eps, rev, p.err.Error())
+			case rpctypes.ErrCompacted:
+				plog.Errorf("%s cannot check the hash of peer(%q) at revision %d: local node is lagging behind(%q)", s.ID(), p.eps, rev, p.err.Error())
+			}
+		}
+	}
+	if mismatch > 0 {
+		return fmt.Errorf("%s is corrupt", s.ID())
```
Comment (xiang90) [inline etcdserver/corrupt.go]: this is not exactly true. maybe other node is corrupted. all we know is that there is a state mismatch inside the cluster.

File: etcdserver/corrupt.go (hunk 22 above)
Comment (xiang90) [reply etcdserver/corrupt.go]: if there is no snapshot, not checking the hash is fine for now.

File: embed/etcd.go (hunk 23 above)
Comment (gyuho) [reply embed/etcd.go]: We still need this to close peer listeners after `startPeerListeners`

File: etcdserver/corrupt.go (hunk 24 above)
Comment (gyuho) [reply etcdserver/corrupt.go]: You are right. How about `%s found data inconsistency with peers`?

File: etcdserver/corrupt.go (hunk 22 above)
Comment (gyuho) [reply etcdserver/corrupt.go]: Ok, I will assume `RaftCluster` is populated, and change the e2e test snapshot count.

File: etcdserver/corrupt.go (hunk 24 above)
Comment (xiang90) [reply etcdserver/corrupt.go]: sure.

File: etcdserver/config.go [hunk 25]
```diff
@@ -66,7 +66,8 @@ type ServerConfig struct {
 
 	AuthToken string
 
-	CorruptCheckTime time.Duration
+	InitialCorruptCheck bool
```
Comment (xiang90) [inline etcdserver/config.go]: add doc string?

File: embed/etcd.go (hunk 23 above)
Comment (xiang90) [reply embed/etcd.go]: what is the problem to return a non-nil server?

File: etcdserver/config.go (hunk 25 above)
Comment (gyuho) [reply etcdserver/config.go]: Fixed.

File: embed/etcd.go (hunk 23 above)
Comment (gyuho) [reply embed/etcd.go]: `e.Server` is not nil, then it will block forever inside `defer` with `EtcdServer.Close`, since the server never got started.

**Comment by xiang90** [comment]
LGTM

File: etcdserver/corrupt.go [hunk 26]
```diff
@@ -16,14 +16,56 @@ package etcdserver
...
+	h, rev, _, err := s.kv.HashByRev(0)
+	if err != nil {
+		return fmt.Errorf("%s failed to fetch hash (%v)", s.ID(), err)
+	}
+	peers := s.getPeerHashKVs(rev)
+	mismatch := 0
+	for _, p := range peers {
+		if p.resp != nil {
+			peerID := types.ID(p.resp.Header.MemberId)
+			if h != p.resp.Hash {
```
Comment (fanminshi) [inline etcdserver/corrupt.go]: I think we also need to reason about compaction revision. Calling on hashKV(100) on all different members does not guarantee that the hash is same among all member even those there is no corruption.

Suppose local member's highest rev is 100 and has no compaction. then HashKV hashes keys from rev = 0 to rev 100. If one of peer member has compacted at 50, then calling hashKV results HashKV hashes keys from rev= 50 to 100.

Comment (gyuho) [reply etcdserver/corrupt.go]: Yeah, just added compact revision checks to handle that case.

File: etcdserver/corrupt.go [hunk 27]
```diff
@@ -16,14 +16,56 @@ package etcdserver
...
+	h, rev, crev, err := s.kv.HashByRev(0)
+	if err != nil {
+		return fmt.Errorf("%s failed to fetch hash (%v)", s.ID(), err)
+	}
+	peers := s.getPeerHashKVs(rev)
+	mismatch := 0
+	for _, p := range peers {
+		if p.resp != nil {
+			peerID := types.ID(p.resp.Header.MemberId)
+			if h != p.resp.Hash && crev == p.resp.CompactRevision {
```
Comment (xiang90) [inline etcdserver/corrupt.go]: we also need to log that we ignore checking as warning if the crev is not matched.

File: etcdserver/corrupt.go [hunk 28]
```diff
@@ -16,14 +16,56 @@ package etcdserver
...
+				plog.Errorf("%s's hash %d != %s's hash %d (revision %d, peer revision %d, compact revision %d)", s.ID(), h, peerID, p.resp.Hash, rev, p.resp.Header.Revision, crev)
+				mismatch++
+			}
+			continue
+		}
+		if p.err != nil {
+			switch p.err {
+			case rpctypes.ErrFutureRev:
+				plog.Errorf("%s cannot check the hash of peer(%q) at revision %d: peer is lagging behind(%q)", s.ID(), p.eps, rev, p.err.Error())
```
Comment (xiang90) [inline etcdserver/corrupt.go]: warning

File: etcdserver/corrupt.go [hunk 29]
```diff
@@ -16,14 +16,56 @@ package etcdserver
...
+			case rpctypes.ErrCompacted:
+				plog.Errorf("%s cannot check the hash of peer(%q) at revision %d: local node is lagging behind(%q)", s.ID(), p.eps, rev, p.err.Error())
```
Comment (xiang90) [inline etcdserver/corrupt.go]: warning

File: etcdserver/corrupt.go (hunk 27 above)
Comment (gyuho) [reply etcdserver/corrupt.go]: Yeah fixed.

File: etcdserver/corrupt.go [hunk 30]
```diff
@@ -16,14 +16,61 @@ package etcdserver
...
+	mismatch := 0
+	for _, p := range peers {
+		if p.resp != nil {
+			peerID := types.ID(p.resp.Header.MemberId)
+			if h != p.resp.Hash {
+				if crev == p.resp.CompactRevision {
+					plog.Errorf("%s's hash %d != %s's hash %d (revision %d, peer revision %d, compact revision %d)", s.ID(), h, peerID, p.resp.Hash, rev, p.resp.Header.Revision, crev)
+					mismatch++
+				} else {
+					plog.Warningf("%s hash mismatch with peer %s at revision %d (compact revision %d, peer compact revision %d)", s.ID(), peerID, rev, crev, p.resp.CompactRevision)
```
Comment (xiang90) [inline etcdserver/corrupt.go]: this mismatch is expected. we should log this as `cannot check hash since the compact reversion is different`

File: etcdserver/corrupt.go [hunk 31]
```diff
@@ -16,14 +16,61 @@ package etcdserver
...
+	mismatch := 0
+	for _, p := range peers {
+		if p.resp != nil {
+			peerID := types.ID(p.resp.Header.MemberId)
+			if h != p.resp.Hash {
+				if crev == p.resp.CompactRevision {
+					plog.Errorf("%s's hash %d != %s's hash %d (revision %d, peer revision %d, compact revision %d)", s.ID(), h, peerID, p.resp.Hash, rev, p.resp.Header.Revision, crev)
+					mismatch++
+				} else {
+					plog.Warningf("%s cannot check hash since the compact reversion is different at revision %d (compact revision %d, peer %s compact revision %d)", s.ID(), rev, crev, peerID, p.resp.CompactRevision)
```
Comment (xiang90) [inline etcdserver/corrupt.go]: well. i mean

`%s cannot check hash of peer(%s): peer has a different compact revision %d (revision:%d)

**Comment by xiang90** [comment]
lgtm if CI passes
